### PR TITLE
Add gain circle

### PR DESCRIPTION
**Pull request by keikawa**
This is related to #362 and #943.

### CODE REVIEW

File: skrf/network.py [hunk 1]
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
...
+        --------
+        stability_circle
+        max_gain : Maximum available and stable power gain
+        max_stable_gain : Maximum stable power gain
+        unilateral_gain : Mason's unilateral power gain
+        
+
+        """
+
+        if self.nports != 2:
```
Comment (Vinc0110) [inline skrf/network.py:4467]: It good to check this. Maybe also add a comment in the docstring on this requirement for a 2-port network.

File: skrf/network.py [hunk 2]
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
...
+        >>> rf.plotting.plot_smith(s=sgc, smith_r=1)
+        >>> plt.show()
+
+        Slicing the network allows you to specify a frequency
+
+        >>> sgc = ntwk['1GHz'].gain_circle(target_port='source', gain=2.0)
+        >>> rf.plotting.plot_smith(s=sgc, smith_r=1)
+        >>> plt.show()
+
+        References
```
Comment (Vinc0110) [inline skrf/network.py:4452]: Providing references is very good. But they are not referenced in any of the docstring sections above, are they? (Not sure if they have to be referenced...)

File: skrf/network.py [hunk 3]
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
...
+        if target_port == 'source':
+            reflection = self.s[:, 0, 0]
+        elif target_port == 'load':
+            reflection = self.s[:, 1, 1]
+        else:
+            raise ValueError("Invalid target_port. Use 'load' or 'source'.")
+
+        gain_factor = mf.db10_2_mag(gain) * (1 - npy.abs(reflection) ** 2)
+        if gain_factor > 1:
+            warnings.warn("The specified gain is too large to be realized. Please specify a smaller gain.", RuntimeWarning)
```
Comment (Vinc0110) [inline skrf/network.py:4483]: I found it helpful to add `stacklevel=2` as an argument for `warnings.warn()`:
https://docs.python.org/3/library/warnings.html#warnings.warn

**Review by Vinc0110 — APPROVED** [review]
Very nice! Thanks for adding this PR to quickly!

I like the extensive docstrings and the tests. Most of my comments are very minor. The only thing I don't like is the definition of `target_port` with 'load' and 'source'. Any other opinions on this?

**Comment by keikawa** [comment]
Thank you for your kind review and comment.
As you suggested, I fixed the specification of `target_port` and some documentation. 
Additionally, I corrected a bug where the if statement would not execute when `gain_factor` is `ndarray`.

File: skrf/network.py [hunk 4]
```diff
@@ -4283,7 +4283,7 @@ def vswr_active(self, a: npy.ndarray) -> npy.ndarray:
 
     def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
         r"""
-        Returns a complex number of stability circles for a given port ('load' or 'source').
+        Returns stability circles for a given port ('load' or 'source').
```
[the remark this answers is not in the document]
Comment (keikawa) [reply skrf/network.py]: I changed the specification of `target_port`.

File: skrf/network.py [hunk 5]
```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
         sc = sc_real + 1j * sc_imag
         return sc.T
 
+    def gain_circle(self, target_port: str, gain: float, npoints: int = 181) -> npy.ndarray:
```
[the remark this answers is not in the document]
Comment (keikawa) [reply skrf/network.py]: Similar to `stability_circle`

File: skrf/network.py (hunk 1 above)
Comment (keikawa) [reply skrf/network.py:4467]: I added comment in the docstring.

File: skrf/network.py (hunk 2 above)
Comment (keikawa) [reply skrf/network.py:4453]: I added reference numbers in the docstring section.

File: skrf/network.py (hunk 3 above)
Comment (keikawa) [reply skrf/network.py]: I added `stacklevel=2` argument.

**Review by Vinc0110 — APPROVED** [review]
Very good. Thanks for the changes!

**Review by jhillairet — APPROVED** [review]

**Pull request merged**